tailnet/service_test.go

```diff
@@ -213,3 +226,146 @@ func TestNetworkTelemetryBatcher(t *testing.T) {
 	require.Equal(t, "5", string(batch[0].Id))
 	require.Equal(t, "6", string(batch[1].Id))
 }
+
+func TestWorkspaceUpdates(t *testing.T) {
+	t.Parallel()
+
+	fCoord := tailnettest.NewFakeCoordinator()
+	var coord tailnet.Coordinator = fCoord
+	coordPtr := atomic.Pointer[tailnet.Coordinator]{}
+	coordPtr.Store(&coord)
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+
+	updatesCh := make(chan *proto.WorkspaceUpdate, 1)
+	updatesProvider := &fakeUpdatesProvider{ch: updatesCh}
+
+	uut, err := tailnet.NewClientService(tailnet.ClientServiceOptions{
+		Logger:                   logger,
+		CoordPtr:                 &coordPtr,
+		WorkspaceUpdatesProvider: updatesProvider,
+	})
+	require.NoError(t, err)
+
+	ctx := testutil.Context(t, testutil.WaitShort)
+	c, s := net.Pipe()
+	defer c.Close()
+	defer s.Close()
+	clientID := uuid.New()
+	errCh := make(chan error, 1)
+	go func() {
+		err := uut.ServeClient(ctx, "2.0", s, tailnet.StreamID{
+			Name: "client",
+			ID:   clientID,
+			Auth: tailnet.ClientUserCoordinateeAuth{
+				Auth: &fakeTunnelAuth{},
+			},
+		})
+		t.Logf("ServeClient returned; err=%v", err)
+		errCh <- err
+	}()
+
+	client, err := tailnet.NewDRPCClient(c, logger)
+	require.NoError(t, err)
+
+	// Coordinate
+	stream, err := client.Coordinate(ctx)
+	require.NoError(t, err)
+	defer stream.Close()
+
+	err = stream.Send(&proto.CoordinateRequest{
+		UpdateSelf: &proto.CoordinateRequest_UpdateSelf{Node: &proto.Node{PreferredDerp: 11}},
+	})
+	require.NoError(t, err)
+
+	call := testutil.RequireRecvCtx(ctx, t, fCoord.CoordinateCalls)
+	require.NotNil(t, call)
+	require.Equal(t, call.ID, clientID)
+	require.Equal(t, call.Name, "client")
+	req := testutil.RequireRecvCtx(ctx, t, call.Reqs)
+	require.Equal(t, int32(11), req.GetUpdateSelf().GetNode().GetPreferredDerp())
+
+	// Authorize uses `ClientUserCoordinateeAuth`
+	agentID := uuid.New()
+	agentID[0] = 1
+	require.NoError(t, call.Auth.Authorize(ctx, &proto.CoordinateRequest{
+		AddTunnel: &proto.CoordinateRequest_Tunnel{Id: tailnet.UUIDToByteSlice(agentID)},
+	}))
+	agentID2 := uuid.New()
+	agentID2[0] = 2
+	require.Error(t, call.Auth.Authorize(ctx, &proto.CoordinateRequest{
+		AddTunnel: &proto.CoordinateRequest_Tunnel{Id: tailnet.UUIDToByteSlice(agentID2)},
+	}))
+
+	// Workspace updates
```

Comment:
I feel like this is 2 tests --- one of `ClientUserCoordinateeAuth` up to here, and a second test of `WorkspaceUpdates`